docs/source/receive-adapter.md

```diff
@@ -0,0 +1,156 @@
+# Receive adapter
+
+A _receive adapter_ is a data-plane component which either
+
+- produces [CloudEvents](https://github.com/cloudevents/spec),
+- or imports events and convert them to
+  [CloudEvents](https://github.com/cloudevents/spec),
+
+and forwards those events to one or several consumers (aka _sink_),
+depending upon whether it supports multiple resources (see below for more information).
+
+## Implementating receive adapter
+
+Knative Eventing provides a library (refer later as *the library*) for bootstrapping receive adapter implementations.
+
+The main code should be something like this:
+
+```go
+package main
+
+import (
+  "knative.dev/eventing/pkg/adapter/youradapter"
+  "knative.dev/eventing/pkg/adapter/v2"
+)
+
+func main() {
+  adapter.Main("pingsource",
+    youradapter.NewEnvConfig,
+    youradapter.NewAdapter)
+}
+```
+
+`adapter.Main` is responsible for:
+
+- parsing the command flags and initializing the environment configuration
+returned by `youradapter.NewEnvConfig`,
+- setting up the logger,
+- setting up the profiler server (when enabled),
+- setting up the metrics server,
+- setting up tracing,
+- setting up a ConfigMap watcher (when enabled),
+- setting up a CloudEvent client, targeting a single sink, for single resource adapter,
+or otherwise targeting no sink,
+- and finally running *your adapter* in leader election mode (when enabled) for being
+highly-available and scalable.
+
+## Push vs pull models
+
+The push model describes the processing mode under which the receive adapter
+reacts to incoming requests or events. The pull model is the opposite, the
+receive adapter does not wait for incoming requests or events but instead it fetches,
+or generates events.
+
+The receive adapter library can be used to implement either type of receive adapters.
+For pull-based receive adapters, the library provides a mechanism for making your adapter
+highly-available and scalable.
+
+## Handling multiple resources (aka multi-tenancy)
+
+### Overview
+
+Single-resource receive adapters handle one source instance (CR) at a time, i.e.
+there is a one-to-one mapping between the receive adapter, and the source
+instance specification. Multi-resource receive adapters can handle more than one
+source instance at a time, either all CRs in one namespace or all CRs in the
+cluster.
+
+While single-resource receive adapters work well for high volume, always "on"
+event sources, it is not optimal (i.e. waste precious resources, cpu and memory)
+for sources producing "few" events, such as `PingSource`, `APIServerSource` and
+`GitHubSource`, or for small applications. On the other hand, single-resource
+receive adapters are fairly easy to implement and more transparent compare to
+their multi-resources equivalent.
+
+The multiple resources feature can be enabled to both pull-based and push-based
```

Comment:
Small nit, at first pass I read this as 'multiple resources mean you can combine pull and push models into the same RA' which I don't think is what you mean, could we clarify that?

Reply:
It should be clearer now. Let me know.